rust/dialog-storage/src/storage/backend/s3/access.rs

```diff
@@ -0,0 +1,705 @@
+use web_time::{SystemTime, web::SystemTimeExt};
+
+// ============================================================================
+// Public Types
+// ============================================================================
+
+/// AWS S3 credentials for signing requests.
+#[derive(Debug, Clone)]
+pub struct Credentials {
+    /// AWS Access Key ID
+    pub access_key_id: String,
+    /// AWS Secret Access Key
+    pub secret_access_key: String,
+    /// Optional AWS session token (for temporary credentials)
+    pub session_token: Option<String>,
+}
+
+/// Cloud storage service configuration for request signing.
+///
+/// Contains the signing parameters for different cloud storage services.
+/// Does NOT include bucket, endpoint, or session duration - those are
+/// specified when creating a session.
+#[derive(Debug, Clone)]
+pub enum Service {
+    /// AWS S3 or S3-compatible service (including Cloudflare R2).
+    S3 {
+        /// AWS region (e.g., "us-east-1", "auto" for R2)
+        region: String,
+    },
+}
+
+impl Service {
+    /// Create S3 service configuration.
+    pub fn s3(region: impl Into<String>) -> Self {
+        Self::S3 {
+            region: region.into(),
+        }
+    }
+
+    /// Get the region for this service.
+    fn region(&self) -> &str {
+        match self {
+            Self::S3 { region } => region,
+        }
+    }
+
+    /// Get the service name for signing.
+    fn name(&self) -> &str {
+        match self {
+            Self::S3 { .. } => "s3",
+        }
+    }
+}
+
+/// S3 Access Control List (ACL) settings.
+///
+/// These are canned ACLs supported by S3 and S3-compatible services.
+/// See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html#canned-acl
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Acl {
+    /// Owner gets FULL_CONTROL. No one else has access rights.
+    Private,
+    /// Owner gets FULL_CONTROL. The AllUsers group gets READ access.
+    PublicRead,
+    /// Owner gets FULL_CONTROL. The AllUsers group gets READ and WRITE access.
+    PublicReadWrite,
+    /// Owner gets FULL_CONTROL. The AuthenticatedUsers group gets READ access.
+    AuthenticatedRead,
+    /// Object owner gets FULL_CONTROL. Bucket owner gets READ access.
+    BucketOwnerRead,
+    /// Both the object owner and the bucket owner get FULL_CONTROL.
+    BucketOwnerFullControl,
+}
+
+impl Acl {
+    /// Get the S3 ACL header value.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            Self::Private => "private",
+            Self::PublicRead => "public-read",
+            Self::PublicReadWrite => "public-read-write",
+            Self::AuthenticatedRead => "authenticated-read",
+            Self::BucketOwnerRead => "bucket-owner-read",
+            Self::BucketOwnerFullControl => "bucket-owner-full-control",
+        }
+    }
+}
+
+/// Pre-derived signing key for request signing.
+#[derive(Debug, Clone)]
+struct SessionKey(Vec<u8>);
```

Comment:
Is there coverage for this? (direct or indirect)

Reply:
Yes it is how all of the authorization system works and struct is private because it's only used by the Authority::authorize which is tested in this module.

More end to end tests that verify that expected signatures come out are in the s3 tests that check exact URLs. And we also run s3s test server which actually performs validation of pre-signed URLs.